packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -121,24 +128,65 @@ contract L2ToL2CrossDomainMessenger is ISemver, TransientReentrancyAware {
     /// @param _destination Chain ID of the destination chain.
     /// @param _target      Target contract or wallet address.
     /// @param _message     Message payload to call target with.
-    /// @return The hash of the message being sent, used to track whether the message has successfully been relayed.
-    function sendMessage(uint256 _destination, address _target, bytes calldata _message) external returns (bytes32) {
+    /// @return messageHash_ The hash of the message being sent, used to track whether the message
+    ///                      has successfully been relayed.
+    function sendMessage(
+        uint256 _destination,
+        address _target,
+        bytes calldata _message
+    )
+        external
+        returns (bytes32 messageHash_)
+    {
         if (_destination == block.chainid) revert MessageDestinationSameChain();
         if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert MessageTargetL2ToL2CrossDomainMessenger();
 
         uint256 nonce = messageNonce();
-        emit SentMessage(_destination, _target, nonce, msg.sender, _message);
+        messageHash_ = Hashing.hashL2toL2CrossDomainMessage({
+            _destination: _destination,
+            _source: block.chainid,
+            _nonce: nonce,
+            _sender: msg.sender,
+            _target: _target,
+            _message: _message
+        });
 
+        sentMessages[messageHash_] = true;
         msgNonce++;
 
-        return Hashing.hashL2toL2CrossDomainMessage({
+        emit SentMessage(_destination, _target, nonce, msg.sender, _message);
+    }
+
+    /// @notice Re-emits a previously sent message event for old messages that haven't been
+    ///         relayed yet, allowing offchain infrastructure to pick them up and relay them.
+    /// @dev    Emitting a message that has already been relayed will have no effect, as it is only
+    ///         relayed once on the destination chain.
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _nonce Nonce of the message sent
+    /// @param _sender Address that sent the message
+    /// @param _target Target contract or wallet address.
+    /// @param _message Message payload to call target with.
+    function resendMessage(
+        uint256 _destination,
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        bytes calldata _message
+    )
+        external
+    {
+        bytes32 messageHash = Hashing.hashL2toL2CrossDomainMessage({
             _destination: _destination,
             _source: block.chainid,
-            _nonce: nonce,
-            _sender: msg.sender,
+            _nonce: _nonce,
+            _sender: _sender,
             _target: _target,
             _message: _message
         });
+
+        if (!sentMessages[messageHash]) revert InvalidMessage();
+
+        emit SentMessage(_destination, _target, _nonce, _sender, _message);
```

Comment:
Do you think we should emit a second event like `emit RelayedMessage(...)` to help identify when the `SentMessage` was the original vs. a resend? Wondering if there is some DoS-type attack where you keep re-emitting the same `SentMessage` and cause confusion on e.g. indexers? 

This also makes me wonder if there is any consideration for op-supervisor to be able to handle duplicate `SentMessage` events, since previously they were unique?

Reply:
Would need to ask @axelKingsley @protolambda about the db schema but I think it should be fine. It needs to index every event, meaning its possible to create events that are the same in a different contract, because the user can define the events.

I think its actually better to have a 1:1 mapping where the event is actually exactly the same, so its 1:1, the same event will always result in the same call on the remote side. Adding in a new codepath just seems to add complexity

Reply:
from @axelKingsley 

> Anyway -- no, a duplicate log event won't break the db. Database contents are indexed by Block Number and Log Index, and never by log event contents. BlockNumber+LogIndex is never reused for a given chain. If a contract emitted several "duplicate log events", they would take their own rows in the database.
From my perspective there is no such thing as a Duplicate Log Event, because log events are positional elements inside a block. So I assume you mean duplicate log content.